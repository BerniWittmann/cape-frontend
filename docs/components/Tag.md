# Tag

A Component to view a single Tag

## Props

<!-- @vuese:Tag:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|tag|The Tag object|`Object`|`true`|-|
|closable|Option whether the tag should be deleteable or not|`Boolean`|`false`|false|
|size|Size of the Tag|`'medium'` / `'small'` / `'mini'`|`false`|-|

<!-- @vuese:Tag:props:end -->


## Events

<!-- @vuese:Tag:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|close|Emitted when the close button is clicked|-|

<!-- @vuese:Tag:events:end -->


