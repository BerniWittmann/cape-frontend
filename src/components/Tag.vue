<template>
  <el-tag class="tag" type="info" :style="style" :closable="closable" :size="size" @close="emitClose">{{ tag.name }}</el-tag>
</template>

<script>
import { convertHexToRgba } from '@/utils/helpers'

/*
 * @vuese
 * @group Components
 *
 * A Component to view a single Tag
 */
export default {
  name: 'Tag',
  props: {
    // The Tag object
    tag: {
      type: Object,
      required: true
    },

    // Option whether the tag should be deleteable or not
    closable: {
      type: Boolean,
      default: false
    },

    // Size of the Tag
    size: {
      // `'medium'` / `'small'` / `'mini'`
      type: String
    }
  },

  computed: {
    style() {
      if (!this.tag.color) return {}

      return {
        backgroundColor: convertHexToRgba(this.tag.color, 0.1),
        borderColor: convertHexToRgba(this.tag.color, 0.2),
        color: convertHexToRgba(this.tag.color, 1)
      }
    }
  },

  methods: {
    emitClose() {
      // Emitted when the close button is clicked
      this.$emit('close')
    }
  }
}

</script>

<style lang="scss">
.tag {
  margin-right: 5px;
  &:last-of-type {
    margin-right: 0;
  }

  .el-tag__close {
    color: inherit !important;
    &:hover {
      background-color: inherit !important;
      color: #fff;
    }
  }
}
</style>
