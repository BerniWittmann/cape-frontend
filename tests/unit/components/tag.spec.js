import { shallowMount } from '@vue/test-utils'
import { i18n } from '../setupPlugins'

import Tag from '@/components/Tag.vue'

describe('Components', () => {
  describe('Tag', () => {
    let propsData
    let cmp

    beforeEach(() => {
      propsData = {
        tag: {
          id: '142',
          name: 'My Tag',
          color: '#FF0000'
        }
      }
      render()
    })

    function render() {
      cmp = shallowMount(Tag, {
        i18n,
        propsData
      })
    }

    it('renders', () => {
      expect(cmp.html()).toMatchSnapshot()
      const tagComponent = cmp.find('eltag-stub')
      expect(tagComponent.exists())
    })

    it('renders the name', () => {
      expect(cmp.text()).toEqual('My Tag')
    })

    it('renders the color', () => {
      const tagComponent = cmp.find('eltag-stub')
      expect(tagComponent.exists())
      expect(tagComponent.attributes('style')).toEqual('background-color: rgba(255, 0, 0, 0.1); color: rgb(255, 0, 0);')
      expect(tagComponent.attributes('type')).toEqual('info')
    })

    it('can handle an undefined color', () => {
      cmp.setProps({
        tag: {
          id: '123',
          name: 'My Tag'
        }
      })
      const tagComponent = cmp.find('eltag-stub')
      expect(tagComponent.exists())
      expect(tagComponent.attributes('style')).toEqual('')
      expect(tagComponent.attributes('type')).toEqual('info')
    })

    it('passes the size', () => {
      cmp.setProps({
        size: 'mini'
      })
      const tagComponent = cmp.find('eltag-stub')
      expect(tagComponent.exists())
      expect(tagComponent.props('size')).toEqual('mini')
    })

    it('is not closeable by default', () => {
      const tagComponent = cmp.find('eltag-stub')
      expect(tagComponent.exists())
      expect(tagComponent.props('closable')).toEqual(false)
    })

    it('can be closable', () => {
      cmp.setProps({
        closable: true
      })
      const tagComponent = cmp.find('eltag-stub')
      expect(tagComponent.exists())
      expect(tagComponent.props('closable')).toEqual(true)
    })

    it('emits a close event on close', () => {
      cmp.setProps({
        closable: true
      })
      expect(cmp.emitted().close).toBeFalsy()

      cmp.vm.emitClose()

      expect(cmp.emitted().close).toBeTruthy()
    })
  })
})
