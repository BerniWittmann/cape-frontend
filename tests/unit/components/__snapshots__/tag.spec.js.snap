// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components Tag renders 1`] = `<eltag-stub type="info" class="tag" style="background-color: rgba(255, 0, 0, 0.1); color: rgb(255, 0, 0);">My Tag</eltag-stub>`;
